### PR TITLE
Add mitigation in 17.7 for cases where local and remote host disagree on source generated files.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
We're seeing a lot of crashes in the wild where the local VS workspace and the Remote OOP workspace disagree on what trees are present within each other (for example, in https://github.com/dotnet/roslyn/issues/69234).

When looking at these crashes, we commonly see stuff like:

image

indicating that this is an issue primarily with SG docs.  Our theory (which we are still collecting data/dumps on) is that there is either a bug in our workspace in terms of how we generate SG docs, or there is a buggy SG out there which doesn't operate deterministically.  In either event, the two hosts could then see a different set of documents, which would then cause these crashes.

To address this, we've made a larger/riskier change in main here https://github.com/dotnet/roslyn/pull/69917 to make it so that the local VS workspace always defers to the remote workspace for SG contents.  This should ensure that there is no way for these systems to be out of sync.

However, we do not want to port a large/risky change like that back to 17.7.  So, instead we're considering mitigating the issue in 17.7 such that the features that are currently affected make themselves resilient to this scenario.  This is not something we want to be in 17.8, so if this is accepted, we will undo this change there and only take the full fix.  

When this flows into main, we will want to keep the first commit, which corresponds to https://github.com/dotnet/roslyn/pull/69952.

We will not want to keep the second commit.

### CODE REVIEW

File: src/EditorFeatures/Core.Wpf/Peek/PeekableItemSource.cs [hunk 1]
```diff
@@ -137,6 +137,9 @@ public void AugmentPeekSession(IPeekSession session, IList<IPeekableItem> peekab
                             workspace, document.Id, item.SourceSpan.Start, cancellationToken).ConfigureAwait(false))
                     {
                         var text = await document.GetTextAsync(project.Solution, cancellationToken).ConfigureAwait(false);
+                        if (text is null)
+                            continue;
```
Comment (sharwell) [inline src/EditorFeatures/Core.Wpf/Peek/PeekableItemSource.cs:141]: ❓  Can we put this in `CanNavigateToPositionAsync` instead?

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ISolutionExtensions.cs [hunk 2]
```diff
@@ -48,8 +50,48 @@ public static Document GetRequiredDocument(this Solution solution, DocumentId do
             => solution.GetDocument(documentId) ?? throw CreateDocumentNotFoundException();
 
 #if !CODE_STYLE
-        public static async ValueTask<Document> GetRequiredDocumentAsync(this Solution solution, DocumentId documentId, bool includeSourceGenerated = false, CancellationToken cancellationToken = default)
-            => (await solution.GetDocumentAsync(documentId, includeSourceGenerated, cancellationToken).ConfigureAwait(false)) ?? throw CreateDocumentNotFoundException();
+        public static async ValueTask<Document?> GetRequiredDocumentIncludingSourceGeneratedAsync(
+            this Solution solution,
+            DocumentId documentId,
+            bool throwForMissingSourceGenerated = true,
+            CancellationToken cancellationToken = default)
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ISolutionExtensions.cs:57]: 💡 Overloads are preferred in order to make `cancellationToken` required on all paths

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ISolutionExtensions.cs:57]: will do.

**Comment by CyrusNajmabadi** [comment]
Closing out.  Did not get QB approval.

**Pull request closed without merging**